pkg/api/pod/util.go

```diff
@@ -594,39 +594,56 @@ func dropDisabledFields(
 		// For other types of containers, validateContainers will handle them.
 	}
 
-	if !utilfeature.DefaultFeatureGate.Enabled(features.PodLifecycleSleepAction) && !podLifecycleSleepActionInUse(oldPodSpec) {
-		for i := range podSpec.Containers {
-			if podSpec.Containers[i].Lifecycle == nil {
-				continue
-			}
-			if podSpec.Containers[i].Lifecycle.PreStop != nil {
-				podSpec.Containers[i].Lifecycle.PreStop.Sleep = nil
-			}
-			if podSpec.Containers[i].Lifecycle.PostStart != nil {
-				podSpec.Containers[i].Lifecycle.PostStart.Sleep = nil
+	dropPodLifecycleSleepAction(podSpec, oldPodSpec)
+}
+
+func dropPodLifecycleSleepAction(podSpec, oldPodSpec *api.PodSpec) {
+	if utilfeature.DefaultFeatureGate.Enabled(features.PodLifecycleSleepAction) || podLifecycleSleepActionInUse(oldPodSpec) {
+		return
+	}
+
+	adjustLifecycle := func(lifecycle *api.Lifecycle) {
+		if lifecycle.PreStop != nil && lifecycle.PreStop.Sleep != nil {
+			lifecycle.PreStop.Sleep = nil
+			if lifecycle.PreStop.Exec == nil && lifecycle.PreStop.HTTPGet == nil && lifecycle.PreStop.TCPSocket == nil {
+				lifecycle.PreStop = nil
 			}
 		}
-		for i := range podSpec.InitContainers {
-			if podSpec.InitContainers[i].Lifecycle == nil {
-				continue
-			}
-			if podSpec.InitContainers[i].Lifecycle.PreStop != nil {
-				podSpec.InitContainers[i].Lifecycle.PreStop.Sleep = nil
-			}
-			if podSpec.InitContainers[i].Lifecycle.PostStart != nil {
-				podSpec.InitContainers[i].Lifecycle.PostStart.Sleep = nil
+		if lifecycle.PostStart != nil && lifecycle.PostStart.Sleep != nil {
```

Comment:
It should not be possible to have more than one set because validation checks it, so you SHOULD be able to just `nil` the whole PresStopPostStart.

Reply:
I'm not sure about this before and still. 
I think that **validation happens after this `drop-disabled` finished** . 

So if we're in a situation the feature is disabled:
Let's say that one crazy user is trying to create a pod with `sleep, exec, httpGet` three handlers. If we just `nil` the whole PresStopPostStart, he won't get any error in validation and the pod will create with no handlers, right?

However, if we only `nil` the `sleep`, and the user trying to create a pod with `sleep, exec` two handlers, it may pass the validation as we `nil` the `sleep` before the validation.

In both case, user will not receive a validation error, which I think he should.

Reply:
My apologies - you are correct!  

staging/src/k8s.io/apiserver/pkg/registry/rest/create.go:

```
    strategy.PrepareForCreate(ctx, obj)
    
    if errs := strategy.Validate(ctx, obj); len(errs) > 0 {    
        return errors.NewInvalid(kind.GroupKind(), objectMeta.GetName(), errs)    
    }    
```